flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/BatchTableEnvironment.scala

```diff
@@ -480,11 +480,20 @@ abstract class BatchTableEnvironment(
       normalizedPlan
     }
 
-    // 4. optimize the physical Flink plan
+    // 4. optimize the logical Flink plan
+    val flinkLogicalOptRuleSet = getFlinkLogicalOptRuleSet
```

Comment:
I'm not sure about adding another step in the optimization phase. You mentioned that "If adding ReduceExpressionRule.CALC to logical opt phase, it will increase the the search time.", do you have numbers for this? Otherwise I would add it to logical optimization phase because this is actually where it belongs.

Reply:
@twalthr You're right, ReduceExpressionRule.CALC should add to logical opt phase, but I tried it and it result to java.lang.OutOfMemoryError when I test `testReduceDeterministicUDF`. I believe the calc rules (`FilterCalcMergeRule, ProjectCalcMergeRule, FilterToCalcRule, ProjectToCalcRule, CalcMergeRule`) interact with each other.
If adding a new phase after, there's only calc node, the search space is quite small than in logical opt phase.
I think adding phase is necessary as the number of rule increasing.